spec/abci++/v2.md

```diff
@@ -0,0 +1,180 @@
+# Tendermint v2 Markdown pseudocode
+
+This adds a single-threaded implementation of ABCI++,
+with no optimization for splitting out verifying the header and verifying the proposal.
+
+### Initialization
+
+```go
+h_p ← 0
+round_p ← 0
+step_p is one of {propose, prevote, precommit}
+decision_p ← nil
+lockedValue_p ← nil
+validValue_p ← nil
+validRound_p ← -1
+```
+
+### StartRound(round)
+
+```go
+function startRound(round) {
+    round_p ← round
+    step_p ← propose
+    if proposer(h_p, round_p) = p {
+        if validValue_p != nil {
+            proposal ← validValue_p
+        } else {
+            txdata ← mempool.GetBlock()
+            // getBlockProposal fills in header
+            unpreparedProposal ← getUnpreparedBlockProposal(txdata)
+            proposal ← ABCI.PrepareProposal(unpreparedProposal)
+        }
+        broadcast ⟨PROPOSAL, h_p, round_p, proposal, validRound_p⟩
+    } else {
+        schedule OnTimeoutPropose(h_p,round_p) to be executed after timeoutPropose(round_p)
+    }
+}
+```
+
+### ReceiveProposal
+
+In the case where the local node is not locked on any round, the following is ran:
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, −1) from proposer(h_p, round_p) while step_p = propose do {
+    if valid(v) ∧ ABCI.ProcessProposal(h_p, v).accept ∧ (lockedRound_p = −1 ∨ lockedValue_p = v) {
+        broadcast ⟨PREVOTE, h_p, round_p, id(v,t)⟩ 
+    } else {
+        broadcast ⟨PREVOTE, h_p, round_p, nil⟩ 
+        // Include any slashing evidence that may be sent in the process proposal response
+        for evidence in ABCI.ProcessProposal(h_p, v).evidence_list {
```

Comment:
what are the assumptions about `ABCI.ProcessProposal(h_p, v).evidence_list`? For instance, if the ABCI endpoint is Byzantine, what guarantees do we have that the code cannot make up the evidence?

Reply:
It can make up evidence. The state machine will have still have to interpret this evidence whenever it lands in a block. This gossip is just specifying that ABCI will trigger this message to gossiped around the network, and potentially included according to the {unspecified} mempool processes.

Reply:
Why evidence is being rebroadcasted here?